walk: refactor and add Walk2 function

jayconrod opened the pull request:
**What type of PR is this?**

> Feature

**What package or component does this PR mostly affect?**

> walk

**What does this PR do? Why is it needed?**

Walk2 is a replacement for Walk. Walk is now deprecated and implemented
in terms of Walk2. We avoid an incompatible change here since packages
outside Gazelle import and use walk.

Walk2 takes a callback function that accepts its arguments and returns
its results via structs, similar to language.Language.GenerateRules.
This allows us to add more arguments and results without incompatible
changes. This also lets the callback function return errors instead of
simply printing and maybe exiting.

This change will enable lazy indexing. The callback function will be
able to return a list of additional directories to visit.

This change includes a major refactoring of the directory caching and
parallel traversal functionality. Instead of storing data in a trie,
we now use a random access cache. This should let us read directories
more easily that weren't prepopulated. The cache population code still
pre-parses build files and respects excluded and ignored patterns.

**Which issues(s) does this PR fix?**

Fixes #1891

**Other notes for review**

This is part 2/4. #2072 is the previous change.

jayconrod commented:
As a bonus, `BenchmarkWalk` is about 40% faster on my machine with this implementation.

jayconrod commented:
Test failures seem like flakes in CI. Looks like I don't have the ability to hit the retry button in Buildkite, but I can poke with an empty commit when we're closer to merging.

fmeum commented on cmd/gazelle/fix-update.go:
```diff
@@ -456,6 +465,10 @@ func runFixUpdate(wd string, cmd command, args []string) (err error) {
 		}
+
+		return walk.Walk2FuncResult{
+			Err: errors.Join(errs...),
```
Just noting that this ties us to Go 1.20+, but we do already have test coverage for the minimum supported version and it appears to be at least that high.

fmeum commented on walk/dirinfo.go:
```diff
@@ -0,0 +1,157 @@
+		}
+	}
+
+	// Call c.get for all directory prefixes. c.get always requires the parent to
```
Comment refers to `c`

jayconrod replied:
Fixed

fmeum commented on walk/cache.go:
```diff
@@ -0,0 +1,73 @@
+	entryMap sync.Map
+}
+
+type cacheEntry struct {
```
Is memory usage of the cache a potential concern? Depending on the padding of Go structs, we could reduce the footprint by storing only one of info and err (via two different entry types).

fmeum reviewed (commented):
I've only reviewed half of the commit yet, but I already like what I'm seeing.

fmeum commented on walk/config.go:
```diff
@@ -119,22 +133,38 @@ func (*Configurer) KnownDirectives() []string {
+		// the caller of this method (configure) sets c.Exts[walkNameCache] to the
+		// preprocessed configuration. We copy it to c.Exts[walkName] instead of
+		// re-processing directives.
+		c.Exts[walkName] = c.Exts[walkNameCached]
```
Could we delete the `walkNameCached` key here to save memory?

jayconrod replied:
Done, but note this only removes the map entry, an alias to the configuration. It won't save memory unless `c.Exts` is rehashed.

jbedard commented on walk/config.go:
```diff
@@ -119,22 +133,38 @@ func (*Configurer) KnownDirectives() []string {
+		// re-processing directives.
+		c.Exts[walkName] = c.Exts[walkNameCached]
+	} else {
+		// If c.Exts[walkNameCached] was not set, process directives normally.
```
Can you explain when it would not be set? I would think it should have always been set during the fs-walk?

jayconrod replied:
Updated comment to indicate this branch is only needed for tests. I didn't want to require that all unit tests go through `walk.Walk2` to call `populateCache`. But it makes this code a little awkward.

jbedard commented on walk/walk.go:
```diff
@@ -109,213 +110,379 @@ type WalkFunc func(dir, rel string, c *config.Config, update bool, f *rule.File,
+			Config:       c,
+			Update:       update,
+			File:         info.file,
+			Subdirs:      subdirs,
```
The `subdirs`s was not modified based on `containedByParent` like `regularFiles` was

jayconrod replied:
Fixed, but please double check

jbedard commented on walk/dirinfo.go:
```diff
@@ -0,0 +1,157 @@
+// populateCache loads directory information in a parallel tree traversal.
+// This has no semantic effect but should speed up I/O.
+//
+// populateCache should only be called when recursion is enabled. It attempts
```
"attempts"? Does it not always?

jbedard commented:
I think this accomplishes everything https://github.com/bazel-contrib/bazel-gazelle/pull/1928 was aiming for? Mainly making BUILD load+parse parse of the concurrent fs traversal, that allowing the `excludes` to be part of the fs-traversal as well.

jbedard commented on walk/walk.go:
```diff
@@ -350,105 +518,26 @@ func findGenFiles(wc *walkConfig, f *rule.File) []string {
 }
 
-func shouldFollow(wc *walkConfig, dir, rel string, ent fs.DirEntry) bool {
+func resolveFileInfo(wc *walkConfig, dir, rel string, ent fs.DirEntry) fs.DirEntry {
```
Was there a reason this was reverted? Was there an actual bug being fixed (and should I test+fix that before this PR is merged)?

jayconrod replied:
Ah, I didn't intend to revert #2054. This was a rebasing mistake.

Looking at it closely though, I think it is better to go back to the original behavior of returning the resolved `fs.DirEntry`. I've renamed, documented, and added a test.

I believe the old code has a bug. It calls `os.ReadDir` to get a list of entries, then `IsDir()` to check whether something is a subdirectory, then `shouldFollow`. `IsDir` returns false on a symbolic link though, so this classifies a followable symbolic link to a directory as a regular file. We need to resolve the symbolic link first before calling `IsDir`.

(Technically we should call `filepath.EvalSymlinks` too before calling `os.Stat` in case there's a chain of symbolic links. I haven't changed that here though.)

jayconrod commented on walk/dirinfo.go:
```diff
@@ -0,0 +1,157 @@
+type dirInfo struct {
+	// entries holds the contents of the directory. Symbolic links are resolved
+	// or not depending on configuration. Excluded and ignored files are included.
+	entries []fs.DirEntry
```
Good catch, should save some memory.

jayconrod commented:
There's a CI failure here that's kind of strange. I think it implies I've broken something with excludes, but I can't reproduce it locally. I'll try to track it down tomorrow.

jbedard commented on walk/walk.go:
```diff
@@ -109,213 +110,383 @@ type WalkFunc func(dir, rel string, c *config.Config, update bool, f *rule.File,
+	}
+	hasBuildFileError := err != nil
+
+	configure(w.cexts, w.knownDirectives, c, rel, info.file, info.config)
```
I believe previously [`configure` was not called `if containedByParent`](https://github.com/bazel-contrib/bazel-gazelle/blob/a39914bf8d3d1b953ef79661df1b787b677b209a/walk/walk.go#L159-L162)

jayconrod replied:
Fixed

jbedard commented on walk/walk.go:
```diff
@@ -109,213 +110,383 @@ type WalkFunc func(dir, rel string, c *config.Config, update bool, f *rule.File,
+	}
+
+	containedByParent := info.file == nil && wc.updateOnly
+	w.visits[rel] = visitInfo{
```
Shouldn't the `visitInfo` be be saved even if this directory was excluded? I think this should move above the `isExcludedDir(rel)` check?

jayconrod replied:
Yes, good catch.

jayconrod commented on walk/walk.go:
```diff
@@ -109,213 +110,383 @@ type WalkFunc func(dir, rel string, c *config.Config, update bool, f *rule.File,
+			for _, f := range vi.regularFiles {
+				regularFiles = append(regularFiles, path.Join(prefix, f))
+			}
+			for _, f := range vi.subdirs {
```
I'm not sure. That is the behavior at `master`, but it's not documented in `WalkFunc`. It looks like before #1921, we passed in all non-excluded subdirectories, regardless of whether they contain or will contain a build file. So was this an intended side effect?

I don't have a strong preference here either way. I have a slight preference for passing in the list of all subdirectories, since that's a simple contract to explain. Does this make a difference for extensions you work with?

The Go extension supports `testdata` directories that are usually just files (no `BUILD` file, so globbed) but are sometimes importable packages. It tracks whether `GenerateRules` generates stuff in each directory, but it would be simpler to use `Subdirs` as you suggest. If I remember correctly, that's actually why `GenerateRules` is called in post-order.

jbedard replied:
Another way to think of it would be `Subdirs` are directories that `RegularFiles` does not traverse into, and #1921 maintained that.

I think I've never actually used `Subdirs` though so if you have a strong preference feel free to change it.

jayconrod commented:
I haven't yet been able to reproduce the CI failure. It's not always failing on the same job, so there's something non-deterministic here. It looks like the `# gazelle:exclude` directive is sometimes being ignored.

jbedard commented:
I was having a similar issue in https://github.com/bazel-contrib/bazel-gazelle/pull/1928.

I _think_ it had something to do with `# gazelle:exclude` excluding itself such as `# gazelle:exclude .`? Then in that scenario the current directory should not be traverse any further and none of the files should be recorded. Sorry I don't recall the details :/

jbedard commented on walk/dirinfo.go:
```diff
@@ -0,0 +1,151 @@
+	}
+
+	info.config = configureForWalk(parentConfig, rel, info.file)
+
```
I think we need `info.config.isExcludedDir(rel)` here to check if the BUILD we just parsed excludes it's own directory 🤯

See https://github.com/bazel-contrib/bazel-gazelle/pull/1928/files#diff-6b93e57c1bff94796acb4cf6b14ec71e6f903b2895bef3e1ab0eb53bb7e9122eR480-R487

jayconrod replied:
Done

jayconrod commented:
> I think we need info.config.isExcludedDir(rel) here to check if the BUILD we just parsed excludes itself

Added the check you suggested, but there's a similar check in `walker.visit` already, so it shouldn't cause a change in behavior. `TestExcludeSelf` covers it too.

I added some debug prints to investigate the problem in CI. `language/proto/BUILD.bazel` contains `# gazelle:exclude testdata`, but `language/proto/testdata` was not added to excludes for `language/proto` in CI. It *was* added on my machine. So that's weird.

Running `//:gazelle_ci` with the race detector revealed the problem: multiple goroutines were appending to `wc.excludes`, so they could race and overwrite each other. Trimming the cap for `excludes` and `follow` when cloning `walkConfig` should fix this. CI is green now at least!

fmeum commented on walk/config.go:
```diff
@@ -69,6 +69,16 @@ func getWalkConfig(c *config.Config) *walkConfig {
 
+func (wc *walkConfig) clone() *walkConfig {
+	wcCopy := *wc
+	// Trim cap of exclude and follow. We may append to these slices in multiple
```
This is such a neat trick!

jayconrod commented:
@ptxmac Could you please open a new issue and tag me? Please include enough information to reproduce the issue.
